lib/absinthe/relay/node.ex

```diff
@@ -210,19 +230,57 @@ defmodule Absinthe.Relay.Node do
   "No source non-global ID value given"
   ```
   """
-  @spec to_global_id(atom | binary, integer | binary | nil) :: binary | nil
-  def to_global_id(_node_type, nil) do
-    nil
+  @spec to_global_id(atom | binary, integer | binary | nil, Absinthe.Schema.t | nil) :: {:ok, global_id_t | nil} | {:error, binary}
+  def to_global_id(node_type, source_id, schema \\ nil)
+  def to_global_id(_node_type, nil, _schema) do
+    {:ok, nil}
   end
-  def to_global_id(node_type, source_id) when is_binary(node_type) do
-    "#{node_type}:#{source_id}" |> Base.encode64
+  def to_global_id(node_type, source_id, schema) when is_binary(node_type) do
+    translate_global_id(schema, :to_global_id, [node_type, source_id])
   end
-  def to_global_id(node_type, source_id, schema) when is_atom(node_type) do
+  def to_global_id(node_type, source_id, schema) when is_atom(node_type) and not is_nil(schema) do
     case Absinthe.Schema.lookup_type(schema, node_type) do
       nil ->
-        nil
+        {:ok, nil}
       type ->
-        to_global_id(type.name, source_id)
+        to_global_id(type.name, source_id, schema)
+    end
+  end
+  
+  @doc """
+  Similar to `to_global_id/3` but raises `RuntimeError` if fails to encode global ID.
+  """
+  @spec to_global_id!(atom | binary, integer | binary | nil, Absinthe.Schema.t | nil) :: global_id_t | nil
+  def to_global_id!(node_type, source_id, schema \\ nil) do
+    case to_global_id(node_type, source_id, schema) do
+      {:ok, global_id} ->
+        global_id
+      {:error, err} ->
+        raise RuntimeError, message: err
+    end
+  end
+
+  defp translate_global_id(nil, direction, args) do
+    apply(Absinthe.Relay.Node.IDTranslator.Default, direction, args ++ [nil])
+  end
+  defp translate_global_id(schema, direction, args) do
+    (global_id_translator(:env, schema) || global_id_translator(:schema, schema) || Absinthe.Relay.Node.IDTranslator.Default)
```

Comment:
I'm pretty sure that the environment setting should be the fallback for the explicit schema (`use`) setting, not the other way around.